pkg/middlewares/tracing/wrapper.go

```diff
@@ -21,48 +21,99 @@ func Wrap(ctx context.Context, constructor alice.Constructor) alice.Constructor
 		if constructor == nil {
 			return nil, nil
 		}
-		handler, err := constructor(next)
+		// we need to start tracing when middleware (provided by `constructor`) starts
+		// and finish when middleware ends, but before calling `next`
+		// so we need to wrap `next` into TraceFinisher
+		// and call TraceFinisher when middleware processed
+		tracingFinisher := NewTraceFinisher(next)
+		handler, err := constructor(tracingFinisher)
 		if err != nil {
 			return nil, err
 		}
 
 		if tracableHandler, ok := handler.(Tracable); ok {
 			name, spanKind := tracableHandler.GetTracingInformation()
 			log.FromContext(ctx).WithField(log.MiddlewareName, name).Debug("Adding tracing to middleware")
-			return NewWrapper(handler, name, spanKind), nil
+			return NewTraceStarter(handler, name, spanKind, tracingFinisher), nil
 		}
 		return handler, nil
 	}
 }
 
-// NewWrapper returns a http.Handler struct
-func NewWrapper(next http.Handler, name string, spanKind ext.SpanKindEnum) http.Handler {
-	return &Wrapper{
-		next:     next,
-		name:     name,
-		spanKind: spanKind,
+// NewTraceStarter returns a *TracingStarted struct
+func NewTraceStarter(middleware http.Handler, name string, spanKind ext.SpanKindEnum, finisher *TraceFinisher) *TraceStarter {
+	return &TraceStarter{
+		middleware: middleware,
+		name:       name,
+		spanKind:   spanKind,
+		finisher:   finisher,
 	}
 }
 
-// Wrapper is used to wrap http handler middleware.
-type Wrapper struct {
-	next     http.Handler
-	name     string
-	spanKind ext.SpanKindEnum
+// TraceStarter is used to start tracing for provided middleware and setup TraceFinisher to finish tracing
+type TraceStarter struct {
+	middleware http.Handler
+	name       string
+	spanKind   ext.SpanKindEnum
+	finisher   *TraceFinisher
+	finishFn   func()
+	finished   bool
 }
 
-func (w *Wrapper) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
-	_, err := tracing.FromContext(req.Context())
+func (ts *TraceStarter) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	parentContext := req.Context()
+	_, err := tracing.FromContext(parentContext)
 	if err != nil {
-		w.next.ServeHTTP(rw, req)
+		if ts.middleware != nil {
+			ts.middleware.ServeHTTP(rw, req)
+		}
+		return
+	}
+
+	_, req, ts.finishFn = tracing.StartSpan(req, ts.name, ts.spanKind)
+	defer ts.finish()
+
+	if ts.finisher != nil {
+		ts.finisher.finish = ts.finish
+		ts.finisher.ctx = parentContext
+	}
```

Comment:
instead of having the extra ts.finish method, and the extra ts.finishFn and ts.finished fields, couldn't we simply handle all the logic with:

```
_, req, finishFn = tracing.StartSpan(req, ts.name, ts.spanKind)
if ts.finisher != nil {
    ts.finisher.finish = finishFn
    ts.finisher.ctx = parentContext
} else {
    defer finishFn()
}
````
?